tests/monotouch-test/CoreFoundation/ProxyTest.cs

```diff
@@ -42,5 +44,182 @@ public void Fields ()
 				Dlfcn.dlclose (lib);
 			}
 		}
+#if !__WATCHOS__ && !MONOMAC
+		[Test]
+		public void TestPACParsingScript ()
+		{
+			// get the path for the pac file, try to parse it and ensure that 
+			// our cb was called
+			string pacPath = Path.Combine (NSBundle.MainBundle.BundlePath, "example.pac");
+			NSError error = null;
+			var script = File.ReadAllText (pacPath);
+			var targetUri = new Uri ("http://docs.xamarin.com");
+			var proxies = CFNetwork.ExecuteProxyAutoConfigurationScript (script, targetUri, out error);
+			Assert.IsNull (error, "Null error");
+			Assert.AreEqual (1, proxies.Length, "Length");
+			// assert the data of the proxy, although we are really testing the js used
+			Assert.AreEqual (8080, proxies [0].Port, "Port");
+		}
+
+		[Test]
+		public void TestPACParsingScriptRunLoop ()
+		{
+			// get the path for the pac file, try to parse it and ensure that 
+			// our cb was called
+			CFProxy [] proxies = null;
+			NSError error = null;
+			NSObject cbClient = null;
+
+			string pacPath = Path.Combine (NSBundle.MainBundle.BundlePath, "example.pac");
+
+			var script = File.ReadAllText (pacPath);
+			var targetUri = new Uri ("http://docs.xamarin.com");
+			var runLoop = CFRunLoop.Current;
+
+			// get source, execute run loop in a diff thread
+
+			CFNetwork.CFProxyAutoConfigurationResultCallback cb = delegate (NSObject c, CFProxy [] l, NSError e) {
+				proxies = l;
+				error = e;
+				cbClient = c;
+				runLoop.Stop ();
```

Comment:
What does `runLoop.Stop()` do in this test code?

Reply:
The CFRunLoop is like a never ending while loop that processes all those events that have been added in it until it is stopped. In our case, we want to get the current loop, execute it until our callback is executed. At that point we have all the data but the loop is sitting for more events to be pumped it to be processed (via the add CFRunLoopSource method). Of course, we are done with the events, we know that as a fact, but the runloop does not know, so within the callback, once we have all what we needed, we tell the CFRunLoop to exit the *'while loop'*.

Reply:
Ok. So the CFRunLoop sits there we do stuff in it until the callback calls stop on it - it is kind of a resource that the dev must manually manage since it is not smart enough to stop itself :)

Reply:
Correct.